presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RuntimeReorderJoinSides.java

```diff
@@ -0,0 +1,163 @@
+                List<VariableReferenceExpression> buildJoinVariables = swapped.getCriteria().stream()
+                        .map(JoinNode.EquiJoinClause::getRight)
+                        .collect(toImmutableList());
+                PlanNode newRight = swapped.getRight();
+                if (needLocalExchange(swapped.getRight(), Optional.of(ImmutableSet.copyOf(buildJoinVariables)), context)) {
+                    if (getTaskConcurrency(context.getSession()) > 1) {
+                        newRight = systemPartitionedExchange(
+                                context.getIdAllocator().getNextId(),
+                                LOCAL,
+                                swapped.getRight(),
+                                buildJoinVariables,
+                                Optional.empty());
+                    }
+                    else {
+                        newRight = gatheringExchange(context.getIdAllocator().getNextId(), LOCAL, swapped.getRight());
+                    }
+                }
+
+                JoinNode newJoinNode = new JoinNode(
+                        swapped.getId(),
+                        swapped.getType(),
+                        newLeft,
+                        newRight,
+                        swapped.getCriteria(),
+                        swapped.getOutputVariables(),
+                        swapped.getFilter(),
+                        leftHashVariable,
+                        swapped.getRightHashVariable(),
+                        swapped.getDistributionType());
+
+                log.info("Probe size: " + leftOutputSizeInBytes + " is smaller than Build size: " + rightOutputSizeInBytes + " => invoke join swapping.");
+                return Result.ofPlanNode(newJoinNode);
+            }
+            else {
+                return Result.empty();
+            }
+        }
+        catch (Exception e) {
+            log.error(format("Runtime join reordering of joinNode (%s) failed, due to (%s).", joinNode.getId(), e.getMessage()));
+            context.getWarningCollector().add(new PrestoWarning(PERFORMANCE_WARNING,
+                    format("Runtime join reordering of joinNode (%s) failed, due to (%s).", joinNode.getId(), e.getMessage())));
+            return Result.empty();
+        }
+    }
+
+    private boolean needLocalExchange(PlanNode root, Optional<Set<VariableReferenceExpression>> partitioningColumns, Context context)
+    {
+        PlanNode actual = context.getLookup().resolve(root);
+        if (actual instanceof ExchangeNode) {
+            if (partitioningColumns.isPresent() && ((ExchangeNode) actual).getPartitioningScheme().getPartitioning().getVariableReferences().containsAll(partitioningColumns.get())) {
+                return false;
+            }
+            // when taskParallelism equals to 1
+            else if (!partitioningColumns.isPresent() && ((ExchangeNode) actual).getType() == ExchangeNode.Type.GATHER) {
```

Comment:
I simplified this part. And it become

```
return partitioningColumns.isPresent() || ((ExchangeNode) actual).getType() != ExchangeNode.Type.GATHER;
```

Shouldn't it be `partitioningColumns.isPresent() && ((ExchangeNode) actual).getType() != ExchangeNode.Type.GATHER;`?

Reply:
If this line is reached and there are `partitioningColumns` required by parent, it basically means the current exchange node's partitioning does not match the requirement from the parent join. And so, no matter if the current one is GATHER or partitioned, we will need to add an exchange. So, I think `||` makes more sense.

Reply:
But if the current is gather, why do we need to repartition it? If the planner decided that the gather exchange is required?

Reply:
A made-up example: Suppose Join's build-side op chain is ... -> [GATHER] -> Sort -> Join. In this case, the join would like partitioned parallel streams for build side (per the TaskParallelism config), but the data is single stream after Sort[/or some other ops requiring single stream]?

Reply:
I don't think it is possible in practice.

@rschlussel what do you think?

Reply:
Ok, if it is not possible to see Ops needing repartition from a single stream (namely single stream ops will always sit around the root of a stage), then I agree using `&&` makes more sense.